torch/fx/passes/reinplace.py

```diff
@@ -0,0 +1,540 @@
+    # (This is asserted by codegen for ATen, but might not be true
+    # for other arbitrary operators).
+    assert len(inplace_overloads_with_matching_schemas) == 1
+    inplace_op = inplace_overloads_with_matching_schemas[0]
+    return inplace_op
+
+def _maybe_get_base_of_curr_node(self_arg):
+    # If the current arg was created from a view op, return the node
+    # corresponding to tensor that it was viewed from.
+    if _is_view_op(self_arg.target):
+        assert isinstance(node.args[0], torch.fx.node.Node)
+        return self_arg.args[0]
+    # We might need to look one level down:
+    # for multi-output view ops like .split(),
+    # The split op creates a list of tensors, and _operator.getitem
+    # is recorded into the FX graph to pass individual tensors from that list
+    # into other operators.
+    if node.target is _operator.getitem:
+        list_arg = self_arg.args[0]
+        if _is_view_op(list_arg.target):
+            assert isinstance(list_arg.args[0], torch.fx.node.Node)
+            return list_arg.args[0]
+    return None
+
+def _can_reuse_tensor_memory(self_arg, *, view_of_self: Optional[torch.fx.node.Node] = None):
+    nodes_used = self_arg.meta['node_usages'] if 'node_usages' in self_arg.meta else []
+    # If the current tensor argument is used as an input to any ops later in the graph,
+    # then we can't re-use its memory
+    if any(_ for n_idx, n in nodes_used if n_idx > idx):
+        return False
+    # Special case; if the current node is base, and we're trying to re-inplace
+    # an operator to v where v = view_op(base, args...),
+    # Then it is safe to re-inplace v, which will directly modify, base,
+    #     **as long as** either:
+    # (1) base is also not used anywhere later in the program
+    # (2) base used exactly once later in the program, in the following operation:
+    #     base_updated = view_inverse_op(base, v_updated, args...)
+    # where "args..." must match up exactly, and "view_inverse_op"
+    # must be the "inverse" of view, according to the inverse rules
+    # in the functionalization pass.
+    maybe_base_of_self = _maybe_get_base_of_curr_node(self_arg)
+    if maybe_base_of_self is not None:
+        return _can_reuse_tensor_memory(base_of_self, view_of_self=self_arg)
+    return True
+
+_VIEW_INVERSE_MAP = {
+    torch.ops.aten.diagonal.default: torch.ops.aten.diagonal_scatter.default,
+    torch.ops.aten.select.int: torch.ops.aten.select_scatter.default,
+    torch.ops.aten.slice.Tensor: torch.ops.aten.slice_scatter.default,
+    torch.ops.aten.as_strided.default: torch.ops.aten.as_strided_scatter.default,
+}
+
+# Uses the 'view_of' metadata set by FunctionalizationMetadataProp
+# to determine if two different nodes are identical views of the same base.
+# For example:
+#    a1 = x.view(2, 4)
+#    a2 = a1[0]
+#    b1 = x.view(2, 4)
+#    b2 = b1[0]
+# Here, b2 and a2 are matching view tensors, because their chain of views is identical.
```

Comment:
if the chain of views is not identical, but the strides are identical, would it make sense for these to count as matching views?

Reply:
I spent some time thinking about this and I'm tempted to save as a follow-up lol. I think you're probably right, and it could simplify this code. It would probably also let us re-inplace any `*_scatter` ops that users use directly in their program, not just from functionalization. I'm a bit worried about false positives though, and making sure we don't accidentally re-inplace something we shouldn't.

Reply:
Ok, I think I have an even simpler version of the "remove scatter ops" logic. Written below:
```
# Given an op that we're trying to re-inplace, "b = foo(a)",
# And given a {view}_scatter op that shows up later in the graph, "y = {view}_scatter(base, x, args...)"
# Then re-inplacing `foo()` would allow us to remove the `{view}_scatter` op entirely, IF:
# (1) "base" is in alias_set(a)
# (2) "x" is in alias_set(b)
```

There's a caveat though: If you have a chain of `{view}_scatter` calls that feed into each other, where the correct output would be to remove all of them, it's not enough to check that `x in alias_set(b)`. "x" might be the output (or an alias of the output) of one of the other `{view}_scatter` ops that you want to remove. Example:
```
  #    b = a[1]
  #    c = b[1]
  #    c_updated = c.add(1)
  #    b_updated = select_scatter(b, c_updated, 0, 1)
  #    a_updated = select_scatter(a, b_updated, 0, 1)
```

I think the high bit here though is that I'm pretty sure it's unnecessary to check for strides or "identical chains of views" here, because it should be enough just to check that the inputs to the scatter ops are in the correct alias sets.

Reply:
Nevermind - it looks like like strides / chains of view metadata are still needed.